src/itwinai/torch/trainer.py

```diff
@@ -1212,42 +1180,420 @@ def distributed(func):
+    },
+    "tune_config": {
+        "num_samples": 1,  # Number of trials to run, increase for more thorough tuning
+        "metric": "loss",
+        "mode": "min",
+    },
+    "run_config": {
+        "checkpoint_at_end": True,  # Save checkpoint at the end of each trial
+        "checkpoint_freq": 10,  # Save checkpoint every 10 iterations
+        "storage_path": "ray_results",  # Directory to save results, logs, and checkpoints
+    },
+    "train_loop_config": {
+        "learning_rate": 1e-3,
+        "batch_size": 32,
+        "epochs": 10,
+        "shuffle_train": False,
+        "shuffle_validation": False,
+        "shuffle_test": False,
+        "pin_gpu_memory": False,
+        "optimizer": "adam",
+        "loss": "cross_entropy",
+        "optim_momentum": 0.9,
+        "optim_weight_decay": 0,
+        "num_workers_dataloader": 4,
+        "random_seed": 21,
+    },
+}
+
+
+class RayTorchTrainer(Trainer):
+    """A trainer class for distributed training and hyperparameter optimization
+    using Ray Train/ Tune and PyTorch.
+
+    Args:
+        config (Dict): A dictionary of configuration settings for the trainer.
+        strategy (Optional[Literal["ddp", "deepspeed"]]):
+            The distributed training strategy to use. Defaults to "ddp".
+        name (Optional[str]): Optional name for the trainer instance. Defaults to None.
+        logger (Optional[Logger]): Optional logger instance. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        config: Dict,
+        strategy: Optional[Literal["ddp", "deepspeed"]] = "ddp",
+        name: Optional[str] = None,
+        logger: Optional[Logger] = None,
+    ) -> None:
+        super().__init__(name=name)
+        self.logger = logger
+        self._set_strategy_and_init_ray(strategy)
+        self._set_configs(config=config)
+
+    def _set_strategy_and_init_ray(self, strategy: str):
+        """Set the distributed training strategy. This will initialize the ray backend.
+
+        Args:
+            strategy (str): The strategy to use for distributed training.
+                Must be one of ["ddp", "deepspeed"].
+
+        Raises:
+            ValueError: If an unsupported strategy is provided.
+        """
+        if strategy == "ddp":
+            self.strategy = RayDDPStrategy()
+        elif strategy == "deepspeed":
+            self.strategy = RayDeepSpeedStrategy()
+        else:
+            raise ValueError(f"Unsupported strategy: {strategy}")
+
+    def _set_configs(self, config: Dict):
+        self.config = deep_update(DEFAULT_CONFIG, config)
+        self._set_scaling_config()
+        self._set_tune_config()
+        self._set_run_config()
+        self._set_train_loop_config()
+
+    @property
+    def device(self) -> str:
+        """Get the current device from distributed strategy.
+        Returns:
+            str: Device string (e.g., "cuda:0").
+        """
+        return self.strategy.device()
+
+    def create_dataloaders(
```

Comment:
Who is calling this method? Is this used?

Reply:
Yes this is the same as for the TorchTrainer, it might be used if a use case is fine using the default implementation. If we change the inheritance it could of course be removed as it is the same as the `create_dataloaders` function of the TorchTrainer